core/node/node_framework/examples/main_node.rs

```diff
@@ -158,7 +159,16 @@ fn main() -> anyhow::Result<()> {
         .add_proof_data_handler_layer()?
         .add_healthcheck_layer()?
         .build()
-        .run()?;
-
-    Ok(())
+        .run_with_setup(|pre_run| {
+            Box::pin(async move {
+                // We want to make sure that the healthcheck server is the very first
+                // thing that starts.
+                pre_run.launch_task(HealthCheckTask::name())?;
+
+                // ...imagine that there is some setup work here -- ensuring genesis state, running some migration,
+                // loading some data, etc..
```

Comment:
I think this approach works (other than minor interface nits), but isn't quite declarative. I wonder whether we could do better. One idea that I've come up with is to describe task preconditions (e.g., "initialized node storage") that would essentially lead to tasks being wrapped like this:

```rust
pub trait Task {
    // existing contents

    fn preconditions(&self) -> Preconditions {
        Preconditions::initialized_storage()
    }
    // or you can define `add_task_with_preconditions` in `ServiceContext`
}

struct WrappedTask<T: Task> {
    preconditions: Preconditions,
    inner: task,
}

impl<T: Task> WrappedTask<T> {
    async fn run(self, stop_receiver: StopReceiver) -> anyhow::Result<()> {
        self.preconditions.wait(stop_receiver.clone()).await?; // cancellation handling omitted for brevity
        self.inner.run(stop_receiver).await
    }
}
```

While it's possible to inject preconditions into each task manually, this approach avoids boilerplate and allows to define default preconditions, so that they can be opted *out of* (which would be the case with the healthcheck server, storage recovery etc.), rather than opted *into*.

I'm not sure how preconditions should be organized internally. It's tempting to say that preconditions should be fulfilled by other tasks (to fit them into the resource / task model); not entirely sure this makes sense. E.g., for the "initialized storage" precondition, you might say that storage initialization is a task that checks / performs initialization. In this case, preconditions can be implemented using watch channels. If fulfilled preconditions are declared in the task interface, it would be possible to "statically" (i.e., given a node config) check that all requested preconditions are actually fulfilled by other tasks etc.; but you'd probably need to support finite tasks for this. A simpler option would be to make preconditions a templated resource.

Another view on preconditions in "node lifecycle"; i.e., assigning different tasks to different parts of the node lifecycle ("initialization", "main phase" etc.). Intuitively, this sounds more restrictive and less expressive than the preconditions model.

Reply:
Funnily, that roughly resembles the design I initially came up with and then discarded.
The problem that I see with an opt-out approach is that for the setup you probably want to have a greater control on the situation.
Imagine that two independent groups of developers work on two big features and decide to implement this opt-out pattern as separate tasks (for whatever reason), and both of them involve some kind of breaking change (e.g. running some kind of storage migration implemented in Rust). This can be merged unnoticed -- because you can never see the "setup" code easily, or even suspect that it exists.

This is why at least for now I chose the opt-in approach where we assume that the "node composer" understands the state of the node and can compose a sequence of actions required to alter its state.

Reply:
But at the same time I certainly can relate to your message, I just kinda feel that if there is a tradeoff between explicitness and declarativity, now is the time to choose explicitness.

More opinions are appreciated, cc @RomanBrodetski

Reply:
Well, my proposal could easily be reworked to have opt-in if we remove the default implementation of `Task::preconditions()`. I still think it makes sense because it allows making waiting for first L1 batch in the storage (e.g., a part of API server) to be more declarative. As another option that requires less new notions:

- Add a resource for preconditions. Could be a templated one, but could be not, e.g. we may have `InitializedStorage` resource.
- Provide `async fn wait_until_ready()` as a part of this resource; in case of `InitializedStorage`, it may return general storage info (e.g., optional snapshot recovery metadata).
- Require this resource in the relevant components. Call `wait_until_ready()` at the start of the corresponding task(s).
- Provide this resource using a separate task (`InitializeStorageTask`?) which would encapsulate genesis / snapshot recovery.

I think task definitions could be made more declarative by having a generic wrapper for tasks with preconditions, so that the main task code still doesn't change:

```rust
struct MyTask {
    // ...
}

#[async_trait]
impl Task for MyTask {
    // ...
}

// in wiring code
let task = MyTask { /* ... */ };
let initialized_storage = context.get_resource::<InitializedStorage>()?;
let task = WithPreconditions::new(task, vec![initialized_storage]);
// ^ can probably be made more eloquent via `TaskExt` extension trait:
let task = task.with_precondition(initialized_storage);
context.add_task(task);

struct WithPreconditions<T> {
    task: T,
    preconditions: Vec<Box<dyn Precondition>>,
}

#[async_trait]
trait Precondition: Resource {
    async fn wait_until_ready(&self); // probably doesn't make sense to define output
}
```

Reply:
But then again -- with an opt-in scheme implemented this way one may forget to add a ready storage precondition (e.g. because they're simply not familiar with the framework) when it would really be required for the component to work correctly.

I try to judge from the perspective of easiness to misuse or do something wrong because of a lack of familiarity with the framework. The way I implemented it is kinda "stupid", but it's *understandably* stupid -- you don't need to know the framework to see all the code that runs before the node has started.

It surely comes with a significant disadvantage: you have to write the code for preconditions yourself, and you must be aware of the set of tasks that has been added to the service. I can already see that it may be a problem for a generic runner that would compose the node based on the node spec.

But at the same time, I'm afraid that we may overcomplicate the framework too much for people to give up on using it as intended.